src/services/bunker.py

```diff
@@ -1,15 +1,344 @@
+        )
+
+        logger.info({"msg": f"Normal CL rebase: {normal_cl_rebase} Gwei"})
+        return normal_cl_rebase
+
+    def _calculate_nearest_and_far_cl_rebase(self, cdtp: CommonDataToProcess) -> tuple[Gwei, Gwei]:
+        """
+        Calculate CL rebase from nearest and far epochs to ref epoch given the changes in withdrawal vault
+        """
+        logger.info({"msg": "Calculating nearest and far CL rebase"})
+        nearest_slot = (cdtp.ref_blockstamp.slot_number - NEAREST_EPOCH_DISTANCE * self.c_conf.slots_per_epoch)
+        far_slot = (cdtp.ref_blockstamp.slot_number - FAR_EPOCH_DISTANCE * self.c_conf.slots_per_epoch)
+        nearest_blockstamp = self.cm.get_first_non_missed_slot(cdtp.ref_blockstamp, SlotNumber(nearest_slot))
+        far_blockstamp = self.cm.get_first_non_missed_slot(cdtp.ref_blockstamp, SlotNumber(far_slot))
+        nearest_cl_rebase = self._calculate_cl_rebase_from(cdtp, nearest_blockstamp)
+        far_cl_rebase = self._calculate_cl_rebase_from(cdtp, far_blockstamp)
+
+        logger.info({"msg": f"CL rebase {nearest_cl_rebase,far_cl_rebase=}"})
+        return nearest_cl_rebase, far_cl_rebase
+
+    def _calculate_cl_rebase_from(self, cdtp: CommonDataToProcess, prev_blockstamp: BlockStamp) -> Gwei:
+        """
+        Calculate CL rebase from prev_blockstamp to ref_blockstamp given the changes in withdrawal vault
+        """
+        prev_epoch = prev_blockstamp.slot_number // self.c_conf.slots_per_epoch
+        logger.info({"msg": f"Calculating CL rebase between {prev_epoch,cdtp.ref_epoch=}"})
+
+        ref_lido_balance_with_vault = sum(self._calculate_lido_balance_with_vault(
+            cdtp.ref_lido_validators, cdtp.ref_blockstamp
+        ))
+
+        prev_all_validators = self.w3.cc.get_validators(prev_blockstamp.state_root)
+        prev_lido_validators = self.w3.lido_validators.filter_lido_validators(cdtp.lido_keys, prev_all_validators)
+        prev_lido_balance_with_vault = sum(self._calculate_lido_balance_with_vault(
+            prev_lido_validators, prev_blockstamp
+        ))
+
+        # handle 32 ETH balances of freshly baked validators, who was activated between epochs
+        validators_diff_in_gwei = (len(cdtp.ref_lido_validators) - len(prev_lido_validators)) * MIN_DEPOSIT_AMOUNT
+        assert validators_diff_in_gwei >= 0, "Validators diff should be positive or 0. Something went wrong with CL API"
+
+        corrected_prev_lido_balance_with_vault = (
+            prev_lido_balance_with_vault
+            + validators_diff_in_gwei
+            - self._get_withdrawn_from_vault_between(prev_blockstamp, cdtp.ref_blockstamp)
+        )
+
+        cl_rebase = ref_lido_balance_with_vault - corrected_prev_lido_balance_with_vault
+
+        logger.info({"msg": f"CL rebase from {prev_epoch=}: {cl_rebase} Gwei"})
+
+        return cl_rebase
+
+    def _calculate_lido_balance_with_vault(
+        self, lido_validators: list[LidoValidator], vault_balance_at: BlockStamp
```

Comment:
vault_balance_at - sounds like we expect here address

Reply:
Lets rename somehow

Reply:
done